source/slang/slang-check-decl.cpp

```diff
@@ -6916,6 +6945,111 @@ namespace Slang
+    // 2. a default value: Struct T{}; Struct S {T t;}
+    //    Note, If a type is not default initializable, it doesn't have default value.
+    // it can be associated with default value expression in the constructor signature.
+    // This function helps to check whether either of those 2 conditions are met and create
+    // a default value for the parameter.
+    // It's totally fine that there is no default value for the parameter, in this case, user
+    // code has to provide the argument for this parameter.
+    static Expr* _getParamDefaultValue(SemanticsVisitor* visitor, VarDeclBase* varDecl)
+    {
+        // 1st condition is easy, we can just use the init expression as the default value.
+        if (varDecl->initExpr)
+        {
+            return varDecl->initExpr;
+        }
+
+        // For the 2nd condition, we need to check if the type is default initializable, if so,
+        // we can use the default constructor.
+        return constructDefaultInitExprForVar(visitor, varDecl);
+    }
+
+    void SemanticsDeclBasesVisitor::_synthesizeEmptyConstructor(StructDecl* structDecl)
+    {
+        // If a type already defines any explicit constructors, do not synthesize any constructors.
+        if (_hasExplicitConstructor(structDecl))
+            return;
+
+        // synthesize the signature first.
+        // The constructor's visibility level is the same as the struct itself.
+        // See: https://github.com/shader-slang/slang/blob/master/docs/proposals/004-initialization.md#synthesis-of-constructors-for-member-initialization
+        DeclVisibility ctorVisibility = getDeclVisibility(structDecl);
+
+        // Only the members whose visibility level is higher or equal than the
+        // constructor's visibility level will appear in the constructor's parameter list.
+        List<VarDeclBase*> resultMembers;
+        _searchMembersWithHigherVisibility(structDecl, ctorVisibility, resultMembers);
+
+        // synthesize the constructor signature:
+        // 1. The constructor's name is always `$init`, we create one without parameters now.
+        ConstructorDecl* ctor = _createCtor(this, getASTBuilder(), structDecl);
+        ctor->members.reserve(resultMembers.getCount());
+
+        // 2. Add the parameter list
+        uint32_t paramIndex = 0;
+        bool stopProcessingDefaultValues = false;
+        for (SlangInt i = resultMembers.getCount() - 1; i >= 0; i--)
+        {
+            auto member = resultMembers[i];
+            auto ctorParam = m_astBuilder->create<ParamDecl>();
+            ctorParam->type = (TypeExp)member->type;
+
+            if (!stopProcessingDefaultValues)
+                ctorParam->initExpr = _getParamDefaultValue(this, member);
+
+            if (!ctorParam->initExpr)
+                stopProcessingDefaultValues = true;
+
+            ctorParam->parentDecl = ctor;
+            StringBuilder prefix;
+            prefix << "param_" << paramIndex++ << "_";
```

Comment:
Don't add prefix here, because in the future we may want to support:
```
T v = {member0: expr, member1: expr}
```
that translates to:
```
T v = T(member0: expr, member1: expr)
```
If we added a prefix, this won't work anymore.